src/server/ttl/ttl_checker.rs

```diff
@@ -0,0 +1,279 @@
+}
+
+impl Display for Task {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        match self {
+            Task::UpdatePollInterval(interval) => {
+                write!(f, "[ttl checker] update poll interval to {:?}", interval)
+            }
+        }
+    }
+}
+
+pub struct TTLChecker<E: KvEngine, R: RegionInfoProvider> {
+    engine: E,
+    region_info_provider: R,
+    poll_interval: Duration,
+}
+
+impl<E: KvEngine, R: RegionInfoProvider> TTLChecker<E, R> {
+    pub fn new(engine: E, region_info_provider: R, poll_interval: Duration) -> Self {
+        TTL_CHECKER_POLL_INTERVAL_GAUGE.set(poll_interval.as_millis() as i64);
+        TTLChecker::<E, R> {
+            engine,
+            region_info_provider,
+            poll_interval,
+        }
+    }
+}
+
+impl<E: KvEngine, R: RegionInfoProvider> Runnable for TTLChecker<E, R>
+where
+    E: KvEngine,
+{
+    type Task = Task;
+
+    fn run(&mut self, task: Task) {
+        match task {
+            Task::UpdatePollInterval(interval) => {
+                self.poll_interval = interval;
+                info!(
+                    "ttl checker poll interval is changed to {}s, will be take effect after next round",
+                    interval.as_secs()
+                );
+                TTL_CHECKER_POLL_INTERVAL_GAUGE.set(self.poll_interval.as_millis() as i64);
+            }
+        }
+    }
+}
+
+impl<E: KvEngine, R: RegionInfoProvider> RunnableWithTimer for TTLChecker<E, R> {
+    fn on_timeout(&mut self) {
+        let mut key = vec![];
+        loop {
+            let (tx, rx) = mpsc::channel();
+            if let Err(e) = self.region_info_provider.seek_region(
+                &key,
+                Box::new(move |iter| {
+                    let mut scanned_regions = 0;
+                    let mut start_key = None;
+                    let mut end_key = None;
+                    for info in iter {
+                        if start_key.is_none() {
+                            start_key = Some(info.region.get_start_key().to_owned());
+                        }
+                        TTL_CHECKER_PROCESSED_REGIONS_GAUGE.inc();
+                        scanned_regions += 1;
+                        end_key = Some(info.region.get_end_key().to_vec());
+                        if scanned_regions == 10 {
+                            break;
+                        }
+                    }
+                    if scanned_regions != 0 {
+                        let _ = tx.send(Some((start_key.unwrap(), end_key.unwrap())));
+                    } else {
+                        let _ = tx.send(None);
+                    }
+                }),
+            ) {
+                error!(?e; "ttl checker: failed to get next region information");
+                TTL_CHECKER_ACTIONS_COUNTER_VEC
+                    .with_label_values(&["error"])
+                    .inc();
+                continue;
+            }
+
+            match rx.recv() {
```

Comment:
This will block the whole thread.

Reply:
we do want to do it one by one

Reply:
Using async/await won't change the order.

Reply:
it's in a single thread not sharing with others.